src/pudl/transform/nrelatb.py

```diff
@@ -0,0 +1,300 @@
+
+from typing import Self
+
+import pandas as pd
+from dagster import asset
+from pydantic import BaseModel, field_validator
+
+import pudl.helpers as helpers
+
+IDX_ALL = [
+    "report_year",
+    "mystery_code",
+    "core_metric_case",
+    "core_metric_variable_year",
+    "cost_recovery_period_years",
+    "scenario_atb",
+    "technology_description",
+    "technology_description_detail_1",
+    "technology_description_detail_2",
+    "core_metric_parameter",
+]
+"""Expected primary key columns for the raw nrelatb data.
+
+The normalized core tables we are trying to build will result in a subset of these columns.
+"""
+
+
+class TableUnstacker(BaseModel):
+    """Info needed to unstack a portion of the NREL ATB table."""
+
+    idx: list[str]
+    params: list[str]
+    """Values from the ``core_metric_parameter`` column to be included in this unstack."""
+
+    @field_validator("idx")
+    @classmethod
+    def idx_are_same_or_subset_of_idx_all(cls: Self, idx: list[str]):
+        """Are the :attr:`idx` columns either the same as or a subset of :attr:`IDX_ALL`?"""
+        assert set(idx).issubset(IDX_ALL)
+        return idx
+
+    @property
+    def idx_unstacked(self: Self) -> list[str]:
+        """Primary key columns after the table is unstacked.
+
+        All of the columns in :attr:`idx` except core_metric_parameter.
+        """
+        return [c for c in self.idx if c != "core_metric_parameter"]
+
+
+class Unstacker(BaseModel):
+    """Class that defines how to unstack the :func:`transform_start` table into all of the tidy NREL tables."""
+
+    rate_table: TableUnstacker
+    scenario_table: TableUnstacker
+    tech_detail_table: TableUnstacker
+
+    @property
+    def params_all(self: Self) -> list[str]:
+        """Compilation of all of the parameter values from each of the tables.
+
+        Also check if there are no duplicate params. We expect all of the parameters
+        across the :class:`TableUnstacker` to be unique.
+        """
+        params_all = (
+            self.rate_table.params
+            + self.scenario_table.params
+            + self.tech_detail_table.params
+        )
+        assert not {x for x in params_all if params_all.count(x) > 1}
+        return params_all
+
+
+def unstacker_of_tables() -> Unstacker:
+    """Build an instance of :class:`Unstacker` for the core NREL ATB tables.
+
+    This function is soley for easy access to the reshaping parameters.
+    """
+    unstacker_rate_table = TableUnstacker(
+        idx=[
+            "report_year",
+            "mystery_code",
+            "core_metric_case",
+            "core_metric_variable_year",
+            "cost_recovery_period_years",
+            "technology_description",
+            "core_metric_parameter",
+        ],
+        params=[
+            "inflation rate",
+            "interest during construction - nominal",
+            "tax rate (federal and state)",
+        ],
+    )
```

Comment:
i figured out which primary keys + parameters would result in the least sparse tables by trial and error in a notebook. i could add more how i did that and at least one more defensive check in `transform_unstack` about the nulls in the output columns.

Reply:
this is how i've been swapping params:

```python
for tech_varying_param in Unstacker().tech_detail_table.params:
    try:
        transform_unstack(
            _core_nrelatb__transform_start,
            TableUnstacker(
                idx=Unstacker().scenario_table.idx,
                params=Unstacker().scenario_table.params + [tech_varying_param]
            )
        )
        print(f"WHOAAA you can add {tech_varying_param}")
    except ValueError:
        print(f"don't add {tech_varying_param}")
```